src/inmanta/references.py

```diff
@@ -0,0 +1,149 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+# This file holds reference equivalents for primive types (excluding lists and dicts). Currently they are
+# both a reference and a real value. For a string it is easy to provide a mock value, for a number it is not
+# so much.
+
+import abc
+import typing
+import uuid
+
+import pydantic
+
+# TODO: work out how reference can have secrets as well
+# TODO: can we add the resolving logic and "relations" between the pydantic classes?
+# TODO: should we introduce a distinction between normal values and secret references so that they are never sent back to the server?
+#       this can also be used for binary values for example (like in files)
+
+
+class ValueReferenceModel(pydantic.BaseModel):
+    """A model class that defines the reference to the value. Values are always represented as an object
+    with one or more attributes that can be referenced
+    """
+
+    @classmethod
+    def get_all_reference_types(cls) -> list[type["ValueReferenceModel"]]:
+        subclasses = []
+        for subclass in cls.__subclasses__():
+            subclasses.append(subclass)
+            subclasses.extend(subclass.get_all_reference_types())
+        return subclasses
+
+    ref_id: uuid.UUID = pydantic.Field(default_factory=uuid.uuid4)
+    """ A uuid to be able to refere to this reference.
+    """
+
+    reference_type: str
+    """ A type key that can be used to correctly select the correct pydantic resource
+
+    # TODO: validate that it matches the name of an entity (for consistency)
+    """
+
+    def resolve_reference(self, values: list["ValueReferenceModel"]) -> object:
+        """Resolve the reference.
+
+        :param values: A list of all value reference
+        """
+        print(self.model_fields)
+
+
+class ValueReference(str):
```

Comment:
This is a string so that in the model/compiler you do not have a python object.  https://github.com/inmanta/std/pull/498/files#diff-5e188e9a2f257f5c800131fb03d2c32f7aae6816bfe2ea0ac9812db5a148cffbR646

However, we have to consider to introduce something in the language to safely allow this. For example something like `opaque` which you can also specialize, like `opaque<string>`. This can then both be used for things like allocators and the reference themselves (opaque). These values cannot be assigned to any attribute unless those are opaque as well.

opaque<string> can be assigned to a field of type string. But, the compiler can still guard you from using them where you expect a proper string.

Reply:
> opaque can be assigned to a field of type string. But, the compiler can still guard you from using them where you expect a proper string.

Alternatively, they can not be assigned to a field of type string, but we provide a built-in dereference/unwrap, so you can e.g. `std::print(unwrap(my_opaque))` so you can access it if you really want to but at least it's explicit.

Reply:
for the pure opaque ones yes, but this could also be used for the references like thing where you know that they will produce a string but you cannot use it in the model

Reply:
You should not, but you can, can't you? If we need it for allocation, we need to be able to dereference it in the compiler. Therefore we have to be able to dereference them at compile-time, no matter if it's the model or the plugin domain.

Reply:
But I have to think about it some more. Another thing that popped in my mind would be something like `agent<string>` to indicate that this is a reference that will only be resolvable on the agent. Again, I need to think about it, it might not prove useful.

Reply:
so, that is what I used opaque for:

- opaque: is what you describe and cannot be assigned or accessed. For example for this thing that made you shiver some time ago: https://code.inmanta.com/solutions/athonet/athonet_mpn/-/blob/master/model/services/mpn.cf?ref_type=heads#L35
- opaque<string>: the secret reference attribute that you can assign to password. 

I get the agent<string> idea, but you might also need to resolve it during compilation.